datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -714,193 +912,86 @@ impl PhysicalExpr for InListExpr {
-                        )
-                    }
-                    TimeUnit::Millisecond => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampMillisecond,
-                            TimestampMillisecondArray
-                        )
-                    }
-                    TimeUnit::Microsecond => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampMicrosecond,
-                            TimestampMicrosecondArray
-                        )
-                    }
-                    TimeUnit::Nanosecond => {
-                        make_contains_primitive!(
-                            array,
-                            list_values,
-                            self.negated,
-                            TimestampNanosecond,
-                            TimestampNanosecondArray
-                        )
-                    }
-                },
-                datatype => Result::Err(DataFusionError::NotImplemented(format!(
-                    "InList does not support datatype {:?}.",
-                    datatype
-                ))),
+                _ => self.evaluate_non_dict(value_data_type, array, list_values),
             }
         }
     }
 }
 
+fn dict_to_val_array<K: ArrowDictionaryKeyType>(
+    array: &Arc<dyn Array>,
+) -> Result<Arc<dyn Array>> {
+    let dict_array = array.as_any().downcast_ref::<DictionaryArray<K>>().unwrap();
+    let mut dict_vals = Vec::with_capacity(dict_array.len());
+    for i in 0..array.len() {
+        let (values_array, values_index) = get_dict_value::<K>(&array, i);
+        // Look up value from Index
+        let value = match values_index {
+            Some(values_index) => ScalarValue::try_from_array(values_array, values_index),
+            // Entry was null, so return null
+            None => values_array.data_type().try_into(),
+        }?;
+        dict_vals.push(value);
+    }
+    let vals = ScalarValue::iter_to_array(dict_vals).unwrap();
+    Ok(vals)
+}
+
+// Return a boolean array indicating whether the value is in list_values
+fn evaluate_set(
+    array: &Arc<dyn Array>,
+    list_values: Vec<ColumnarValue>,
+) -> Result<Arc<dyn Array>> {
+    // convert value_list to an array
+    let scalars = list_values
+        .iter()
+        .map(|v| match v {
+            ColumnarValue::Scalar(scalar) => scalar.clone(),
+            ColumnarValue::Array(_array) => {
+                unimplemented!("InList does not yet support nested columns.")
+            }
+        })
+        .collect::<Vec<_>>();
+    let list_array = ScalarValue::iter_to_array(scalars).unwrap();
+
+    let cmp = build_compare(&array, &list_array).unwrap();
```

Comment:
This is known not to handle nulls correctly - https://github.com/apache/arrow-rs/issues/2687